p2p/config/config.go

```diff
@@ -22,6 +32,14 @@ type baseConfig struct {
 	PeerstoreProvider     providers.PeerstoreProvider
 }
 
+type UnicastRouterConfig struct {
+	Logger         *modules.Logger
```

Comment:
This is the first:
1. Package exposed `Logger` in a struct
2. Config without an associated proto

No problems/concerns, but wanted to understand if this is the approach we're taking with submodules?

Reply:
> 1. Package exposed Logger in a struct

The structs in this package are all config structs which are used to pass required parameters to constructor functions (most of which are submodule constructors). They must export their fields in order for consumers to be able to configure the unexported fields of the object under construction, via the respective constructor.

I could understand the argument that it may make more sense for these configs to live in same package as the constructor which they accompany; however, even then, these fields would still need to be exported for package-external consumers to use them (e.g. `p2p` package). The rationale behind the current code organization was to deduplicate `baseConfig#IsValid()` between `RainTreeConfig` and `BackgroundConfig`. To do this without creating an import cycle requires an additional package. At that point, this package seemed most appropriate for `UnicastConfig` to live here as well. I don't have a strong opinion about it, I just felt like this was the most readable and least complex option at the time.

> 2. Config without an associated proto

`RainTreeRouter` and `BackgroundRouter` configs (also defined in this package) do not have an associated protobuf config type, nor do I think it would make sense for them to. In the specific cases of the routers, only one of the required parameters are serializable and it is (appropriately) derived from a `P2PConfig` field (`Address`). This makes it a poor candidate, in my view, to be converted into a protobuf type such that it can be embedded in the P2P config type; however, you've made me aware of the possibility of such a case an potential consistent integration approach. :raised_hands:

```mermaid
mindmap
    root((Submodules))
        (options)
            ASSUMPTIONS
                (MUST be optional, effecting the submodules state and/or behavior)
        (configs)
            (Consideration:
            embedding submodule configs in respective module's protobuf config]
                ASSUMPTIONS
                    (Must be required for a submodule to function properly)
                    ("Multiple submodules may share a common dependency which MAY submodule, 'module-level shared dependencies'")
                        %% ("Initializaiton of such shared depencencies SHOULD be initialized and/or registered (in the case of submodules) in a parent module  in each submodule, ideally via some common and/or embedded implementation.")
                PROS
                    ("Provides clear and consistent convention with respect to how to structure configs in both modules and submodules")
                    ("Supports shared submodules; i.e. such configs could be aggregated with cardinality greater than 1 in one or more modules simultaneously without conflict")
                    ("Encourages usage of dependecy injection for submodules.")
                            ("For example, the P2P module currently sets up a peerstore provider and then passes it to router submdoules via their configs.
                            Since the providers (pstore & height) are submodules, the P2P SHOULD instead register them to the bus and let the routers dendency-inject them.")
                CONS
                    ("Protobuf config can only contain primitive/serializable types")
                        ("Functions and non-submodule, non-serializable dependencies become problematic")
                            ("For example, the P2P module currently sets up a libp2p host and passes it along with a reference to one of its methods to router submdoules via their configs.
                            Forcing submodules to rely only on seralizable configs seems insufficient.")
                ALTERNATIVES
                    (Support either and/or both; embed submodule-specific protobuf configs, where applicable, AND allow for an additional layer of abstraction to support passing high-level objects.]
                        (A protobuf config type MAY be embedded in the submodule's "config type")
                OPEN Qs
                    (Should submodules be constrained in whether they can consider a protobuf config type directly?)
        

```

Reply:
Makes sense to me. My tl;dr takeaway is:

1. **Separate package**: share configs across `RainTreeConfig` and `BackgroundConfig`
2. **No protobuf**: not exposed to the user / node runner.